src/main/host/descriptor/descriptor.c

```diff
@@ -58,22 +69,48 @@ void descriptor_ref(gpointer data) {
     LegacyDescriptor* descriptor = data;
     MAGIC_ASSERT(descriptor);
 
-    (descriptor->referenceCount)++;
-    trace("Descriptor %i ref++ to %i", descriptor->handle,
-          descriptor->referenceCount);
+    (descriptor->refCountStrong)++;
+    trace("Descriptor %i ref++ to %i", descriptor->handle, descriptor->refCountStrong);
 }
 
 void descriptor_unref(gpointer data) {
     LegacyDescriptor* descriptor = data;
     MAGIC_ASSERT(descriptor);
 
-    (descriptor->referenceCount)--;
-    trace("Descriptor %i ref-- to %i", descriptor->handle,
-          descriptor->referenceCount);
+    (descriptor->refCountStrong)--;
+    trace("Descriptor %i ref-- to %i", descriptor->handle, descriptor->refCountStrong);
+
+    utility_assert(descriptor->refCountStrong >= 0);
+    if (descriptor->refCountStrong == 0 && descriptor->refCountWeak == 0) {
+        _descriptor_cleanup(descriptor);
+        _descriptor_free(descriptor);
+    } else if (descriptor->refCountStrong == 0) {
+        // create a temporary weak reference to prevent the _descriptor_cleanup() from calling
+        // descriptor_unrefWeak() and running the _descriptor_free() while still running the
+        // _descriptor_cleanup()
+        descriptor_refWeak(descriptor);
+        _descriptor_cleanup(descriptor);
+        descriptor_unrefWeak(descriptor);
+    }
+}
+
+void descriptor_refWeak(gpointer data) {
+    LegacyDescriptor* descriptor = data;
+    MAGIC_ASSERT(descriptor);
+
+    (descriptor->refCountWeak)++;
+    trace("Descriptor %i weak_ref++ to %i", descriptor->handle, descriptor->refCountWeak);
+}
+
+void descriptor_unrefWeak(gpointer data) {
+    LegacyDescriptor* descriptor = data;
+    MAGIC_ASSERT(descriptor);
+
+    (descriptor->refCountWeak)--;
+    trace("Descriptor %i weak_ref-- to %i", descriptor->handle, descriptor->refCountWeak);
 
-    utility_assert(descriptor->referenceCount >= 0);
-    if(descriptor->referenceCount == 0) {
-        gint handle = descriptor->handle;
+    utility_assert(descriptor->refCountWeak >= 0);
+    if (descriptor->refCountStrong == 0 && descriptor->refCountWeak == 0) {
         _descriptor_free(descriptor);
```

Comment:
It looks like `descriptor_cleanup` is never called if the last reference is a weak ref?

Reply:
`descriptor_cleanup()` should have been called when the last strong reference reached 0.

Reply:
Ah yeah, I'd misread that code. The new structure makes it clearer.